File: backend/app/models/trainings.py
from __future__ import annotations

from datetime import date, datetime

from sqlalchemy import Boolean, Date, DateTime, Float, ForeignKey, Index, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db import Base


class Training(Base):
    __tablename__ = "trainings"
    __table_args__ = (Index("idx_trainings_participant", "participant_open_id"),)

    training_id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    base_record_id: Mapped[str | None] = mapped_column(String, unique=True, nullable=True)
    participant_open_id: Mapped[str] = mapped_column(ForeignKey("members.open_id"), nullable=False)
    name: Mapped[str] = mapped_column(String, nullable=False)
    type: Mapped[str] = mapped_column(String, nullable=False)
    organizer: Mapped[str | None] = mapped_column(String, nullable=True)
    start_date: Mapped[date] = mapped_column(Date, nullable=False)
    end_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    location: Mapped[str | None] = mapped_column(String, nullable=True)
    hours: Mapped[float | None] = mapped_column(Float, nullable=True)
    has_certificate: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    certificate_url: Mapped[str | None] = mapped_column(String, nullable=True)
    reflection: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime,
        default=datetime.utcnow,
        onupdate=datetime.utcnow,
        nullable=False,
    )

    participant: Mapped["Member"] = relationship("Member", back_populates="trainings")

File: frontend/src/api/trainings.ts
import { api } from "./client";
import type { Page } from "../types/api";

export interface Training {
  training_id: number;
  base_record_id?: string | null;
  participant_open_id: string;
  name: string;
  type: string | null;
  organizer: string | null;
  start_date: string | null;
  end_date: string | null;
  location: string | null;
  hours: number | null;
  has_certificate: boolean;
  certificate_url: string | null;
  reflection: string | null;
  created_at?: string;
  updated_at?: string;
}

export const listTrainings = async (params?: {
  participant_open_id?: string;
  type?: string;
  page?: number;
  page_size?: number;
}): Promise<Page<Training>> => {
  const { data } = await api.get<Page<Training>>("/trainings", { params });
  return data;
};

export const getTraining = async (training_id: number | string): Promise<Training> => {
  const { data } = await api.get<Training>(`/trainings/${training_id}`);
  return data;
};
